src/Data/UTxO/Transaction/Cardano/Shelley.hs

```diff
@@ -0,0 +1,373 @@
+    ( SlotNo (..) )
+import Control.Monad
+    ( replicateM )
+import Control.Monad.Fail
+    ( MonadFail )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Short
+    ( toShort )
+import Data.Either.Extra
+    ( eitherToMaybe )
+import Data.UTxO.Transaction
+    ( ErrMkPayment (..), MkPayment (..) )
+import Data.UTxO.Transaction.Cardano.Helpers
+    ( xprvFromBytes )
+import Data.Word
+    ( Word32, Word64 )
+import GHC.Natural
+    ( naturalToInteger )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Cardano.Api.Typed as Cardano
+import qualified Cardano.Chain.Common as Byron
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import qualified Shelley.Spec.Ledger.Address.Bootstrap as Ledger
+import qualified Shelley.Spec.Ledger.Credential as Ledger
+
+-- | Construct a payment 'Init' for /Shelley/ from primitive types.
+--
+-- __examples__:
+--
+-- >>> mkInit Mainnet 430000
+-- >>> mkInit (Testnet (NetworkMagic 1234)) 430000
+--
+-- @since 2.0.0
+mkInit
+    :: NetworkId
+        -- ^ A network tag, Mainnet or Testnet with NetworkMagic specified
+    -> Word64
+        -- ^ A ttl expressed in slot number counted from the beginning of blockchain
+    -> Word64
+        -- ^ fee of tx as taken when constructing change outputs
+        -- see also estimateFee
+    -> Init Shelley
+mkInit net ttl fee = (net, SlotNo ttl, Cardano.Lovelace $ fromIntegral fee)
+
+-- | Estimate a payment fee for /Shelley/ ttransaction using the number of
+-- inputs and outputs and the sorts of witnesses. Please, notice that this
+-- method overetimates the minimum acceptable fees.
+--
+-- __examples__:
+--
+-- >>> estimateFee 1 2 1 0 Cardano.Mainnet
+-- Lovelace 168097
+--
+-- >>> estimateFee 1 2 1 0 (Cardano.Testnet (Cardano.NetworkMagic 12345))
+-- Lovelace 168097
+--
+-- estimateFee 1 2 0 1 Cardano.Mainnet
+-- Lovelace 169725
+-- @since 2.0.0
+estimateFee
+    :: Int
+        -- ^ Number of inputs in tx
+    -> Int
+        -- ^ Number of outputs in tx
+    -> Int
+        -- ^ Number of Shelley witnesses in tx
+    -> Int
+        -- ^ Number of Byron witnesses in tx
+    -> NetworkId
+        -- ^ Network id
+    -> Cardano.Lovelace
+estimateFee nInps nOuts nShelleyWits nByronWits net =
+    Cardano.estimateTransactionFee net minFeeB minFeeA dummyTx nInps nOuts nShelleyWits nByronWits
+  where
+      ttl = SlotNo 10000
+      initialSigData = Cardano.makeShelleyTransaction
+          TxExtraContent
+          { txMetadata = Nothing
+          , txWithdrawals = []
+          , txCertificates = []
+          , txUpdateProposal = Nothing
+          }
+          ttl
+          (Cardano.Lovelace $ naturalToInteger minFeeB)
+          []
+          []
+      dummyTx = Cardano.makeSignedTransaction [] initialSigData
+
+      -- Values to calculate fee : b + a*txsize
+      minFeeA = 44
+      minFeeB = 155381
```

Comment:
These values only hold for current Mainnet (and may even be different tomorrow). I still don't fully understand the need for a _fee estimation_ function since we are constructing the actual transaction and thus, all inputs and outputs are known. There's no need to estimate.

Reply:
I removed feeEstimation - I agree it would be better to service that in external tx one day - https://github.com/input-output-hk/cardano-transactions/pull/17/commits/23deb590347fdd28c2f99a2f008ab42c098cead8